### PR TITLE
Integrate CCS with new search_shards API

### LINKED ISSUE #93730: Search shards API to rewrite query and skip shards
The search shards API is a public API, that the search and async search API use internally when executing cross-cluster search requests (and roundtrips are not minimized). For each remote cluster involved in the search request, search shards will be invoked to collect information about the remote shards that have to be queried. The API accepts a set of indices and returns a set of shards that the coordinating node includes in the search request.  Ideally, such API would accept a query and perform the coordinator can_match phase, so that it does not return shards that could be easily skipped. The coordinator can_match phase skips shards based on index metadata (min and max) that is present in the cluster state, which the CCS coordinating cluster is not aware of for remote indices.

This would help unify the search behaviour between local and remote searches. Today searchable snapshots allow to skip shards completely when the query has a timerange filter against the `@timestamp` field that can not possibly match, and this is good to avoid throwing shards failures for irrelevant shards in the situation where frozen indices are not available (can happen as they don't have replicas).

### PR DESCRIPTION
**Pull request by dnhatn**
This PR integrates CCS with the new search_shards API. With this change, we will be able to skip shards on the coordinator on remote clusters using the timestamps stored in the cluster state.

Relates #94534
Closes #93730

### CODE REVIEW

**Comment by dnhatn** [comment]
run elasticsearch-ci/full-bwc

**Comment by dnhatn** [comment]
run elasticsearch-ci/full-bwc

**Comment by dnhatn** [comment]
run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 1]
```diff
@@ -582,50 +596,97 @@ static SearchResponseMerger createSearchResponseMerger(
...
+                final Transport.Connection connection;
+                try {
+                    connection = remoteClusterService.getConnection(clusterAlias);
+                } catch (NoSuchRemoteClusterException ex) {
+                    // attempt to reconnect but don't wait
+                    remoteClusterService.ensureConnected(clusterAlias, ActionListener.noop());
+                    throw ex;
+                }
+                final String[] indices = entry.getValue().indices();
+                // TODO: support point-in-time
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:653]: I am closing #93730 with a TODO to support point-in-time.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 2]
```diff
@@ -582,50 +596,97 @@ static SearchResponseMerger createSearchResponseMerger(
...
+                        return searchShardsResponses;
                     }
-                );
+                };
+            final StepListener<Void> ensureConnectedListener = new StepListener<>();
+            if (skipUnavailable) {
+                ensureConnectedListener.onResponse(null);
+            } else {
+                remoteClusterService.ensureConnected(clusterAlias, ensureConnectedListener);
+            }
```
Comment (quux00) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:642]: I don't understand what is happening in this if/else block. Reading down into the `remoteClusterService .ensureConnected` code, it looks like the service will check if it is connected to the remote cluster and if so, it calls the onResponse callback of the listener with `null` as you do on line 639. Importantly, it calls listener.onFailure(Exception) if the connection fails.

I get the calling of `listener.onFailure` only if `skipUnavailable==false`, but I don't understand the point of calling `onResponse(null)` on this listener. Especially on line 638 when you know that all you have is an empty StepListener that (I don't think?) is configured to do anything at that point in time.

Is this `onResponse(null)` call necessary? Or is just there for future proofing? I wonder if it needs some code comments to explain why it is being done.

File: server/src/main/java/org/elasticsearch/action/search/SearchShardIterator.java [hunk 3]
```diff
@@ -59,7 +60,9 @@ public SearchShardIterator(
         List<String> targetNodeIds,
         OriginalIndices originalIndices,
         ShardSearchContextId searchContextId,
-        TimeValue searchContextKeepAlive
+        TimeValue searchContextKeepAlive,
+        boolean prefiltered,
+        boolean skip
```
Comment (quux00) [inline server/src/main/java/org/elasticsearch/action/search/SearchShardIterator.java:65]: A minor nice-to-have, but it would be good to move the javadoc from the other constructor down to this one and document the new parameters.

I had to go down to the "getter" methods for `skip` and `prefiltered` to understand what they represented.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:642]: Calling `RemoteClusterService#ensureConnected` can take time as we need to re-establish the connection to the remote cluster if we don't have one. We prefer to fail fast search requests with `skip_unavailable=true` in this case. The detail can be found in https://github.com/elastic/elasticsearch/pull/80589 and https://github.com/elastic/elasticsearch/issues/74773. Happy to walk through if you need more context.

> I don't understand the point of calling `onResponse(null)` on this listener.

`onResponse(null)` means we are skipping `RemoteClusterService#ensureConnected` when `skip_unavailble=true`

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:642]: The logic here is similar to what we have in RemoteClusterAwareClient.

Comment (quux00) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:642]: OK, but if you want to short-circuit should you put a return statement in the `if` block after line 639?

**Review by quux00 — APPROVED** [review]
LGTM. Really nice work here @dnhatn ! Thanks for the help going through the changes with me.

File: server/src/test/java/org/elasticsearch/action/search/TransportSearchActionTests.java [hunk 4]
```diff
@@ -256,100 +257,117 @@ public void testProcessRemoteShards() {
         ) {
             RemoteClusterService service = transportService.getRemoteClusterService();
             assertFalse(service.isCrossClusterSearchEnabled());
```
Comment (quux00) [inline server/src/test/java/org/elasticsearch/action/search/TransportSearchActionTests.java:259]: Should we remove this line? It seems unrelated to what we are testing and is confusing to the reader.

File: server/src/main/java/org/elasticsearch/action/search/SearchShardIterator.java (hunk 3 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/SearchShardIterator.java:77]: ++, I pushed https://github.com/elastic/elasticsearch/pull/95894/commits/bc808bb4100609b4e5d45911e31d7e20819f0503

File: server/src/test/java/org/elasticsearch/action/search/TransportSearchActionTests.java (hunk 4 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/action/search/TransportSearchActionTests.java]: ++, I removed the TransportServices from tests: https://github.com/elastic/elasticsearch/pull/95894/commits/428e098e37a4f81abbcf486912b6bbd4ebc38580

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+import java.util.List;
+import java.util.Optional;
+
+import static org.hamcrest.Matchers.in;
+
+public class CCSCanMatchIT extends AbstractMultiClustersTestCase {
+
+    @Override
+    protected Collection<String> remoteClusterAlias() {
+        return List.of("cluster_a");
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java:53]: ok, see https://github.com/elastic/elasticsearch/pull/95894/commits/f8746263e442b63c45978d22e5ad1e8192c263f7

**Comment by dnhatn** [comment]
@quux00 Thanks for reviews.

File: server/src/main/java/org/elasticsearch/action/search/SearchShardsResponse.java [hunk 6]
```diff
@@ -80,4 +94,22 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(groups, nodes, aliasFilters);
     }
+
+    public static SearchShardsResponse fromLegacyResponse(ClusterSearchShardsResponse oldResp) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchShardsResponse.java:98]: does this need to be public?

File: server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java [hunk 7]
```diff
@@ -142,6 +145,18 @@ public static MockTransportService startTransport(
                     }
                 }
             );
+            newService.registerRequestHandler(
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java:148]: do we need to register both old and new action ?

**Review by javanna — COMMENTED** [review]
Thanks for working on this @dnhatn , I left some comments and questions.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchShardsAction.java [hunk 8]
```diff
@@ -115,6 +115,11 @@ protected void doExecute(Task task, SearchShardsRequest searchShardsRequest, Act
                     indicesAndAliases,
                     concreteIndexNames
                 );
+                if (SearchService.canRewriteToMatchNone(searchRequest.source()) == false) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchShardsAction.java:118]: I opened https://github.com/elastic/elasticsearch/pull/96117 for this.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 9]
```diff
@@ -333,13 +341,16 @@ void executeRequest(
                 } else {
                     AtomicInteger skippedClusters = new AtomicInteger(0);
                     collectSearchShards(
+                        parentTaskId,
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: Thank you for your input! I have reverted the task and added a TODO in https://github.com/elastic/elasticsearch/pull/95894/commits/5656ec8b38cd8680d7a8dac91a6e9bd42ed0b6a5to address the concern. However, I'm not entirely certain if this particular change requires that level of attention.

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                        assertThat("old indices should be prefiltered on coordinator node", "remote_old_index", Matchers.not(in(indices)));
+                    }
+                    connection.sendRequest(requestId, action, request, options);
+                });
+            }
+        }
+        try {
+            SearchSourceBuilder source = new SearchSourceBuilder().query(new RangeQueryBuilder("@timestamp").from(timestamp));
+            SearchRequest request = new SearchRequest("local_*", "cluster_a:remote_*");
+            request.source(source).setCcsMinimizeRoundtrips(randomBoolean());
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java]: Yes, the test was introduced for minimized_round_trips disabled; but the assertions are guaranteed for both cases so we should test both. I adjusted the test to run both cases instead.

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+    }
+
+    public static class ExposingTimestampEnginePlugin extends Plugin implements EnginePlugin {
+
+        @Override
+        public Optional<EngineFactory> getEngineFactory(IndexSettings indexSettings) {
+            if (IndexMetadata.INDEX_BLOCKS_WRITE_SETTING.get(indexSettings.getSettings())) {
+                return Optional.of(EngineWithExposingTimestamp::new);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java:89]: Yes, that's correct.

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+                    }
+                    connection.sendRequest(requestId, action, request, options);
+                });
+            }
+        }
+        try {
+            SearchSourceBuilder source = new SearchSourceBuilder().query(new RangeQueryBuilder("@timestamp").from(timestamp));
+            SearchRequest request = new SearchRequest("local_*", "cluster_a:remote_*");
+            request.source(source).setCcsMinimizeRoundtrips(randomBoolean());
+            ElasticsearchAssertions.assertHitCount(client().search(request).actionGet(), localDocs + remoteDocs);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java]: Great point on asserting skipped shards. I added the assertion in https://github.com/elastic/elasticsearch/pull/95894/commits/bd87504b126f09925233641e728d929b19254353.

>  are the remotely skipped shards accounted in the skipped count returned to the users

Yes, they are.

File: server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java [hunk 13]
```diff
@@ -520,11 +526,8 @@ private GroupShardsIterator<SearchShardIterator> getIterator(
         SearchSourceBuilder source = request.source();
         int i = 0;
         for (SearchShardIterator iter : shardsIts) {
-            if (possibleMatches.get(i++)) {
-                iter.reset();
-            } else {
-                iter.resetAndSkip();
-            }
+            boolean match = possibleMatches.get(i++);
+            iter.reset(match == false);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java]: That's required for the case where all remote shards were skipped and we have to un-skip one of them. I modified this https://github.com/elastic/elasticsearch/pull/95894/commits/1b50a75614393d48fe6eeb5e0ab4874f46c247ea. WDYT?

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 14]
```diff
@@ -582,50 +596,99 @@ static SearchResponseMerger createSearchResponseMerger(
...
-                        Map<String, ClusterSearchShardsResponse> createFinalResponse() {
-                            return searchShardsResponses;
-                        }
+                    @Override
+                    Map<String, SearchShardsResponse> createFinalResponse() {
+                        return searchShardsResponses;
                     }
-                );
+                };
+            final StepListener<Void> ensureConnectedListener = new StepListener<>();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: Yes, please see RemoteClusterAwareClient#doExecute.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 15]
```diff
@@ -582,50 +596,99 @@ static SearchResponseMerger createSearchResponseMerger(
...
+            if (skipUnavailable) {
+                ensureConnectedListener.onResponse(null);
+            } else {
+                remoteClusterService.ensureConnected(clusterAlias, ensureConnectedListener);
+            }
+            ensureConnectedListener.addListener(ActionListener.wrap(nullValue -> {
+                final Transport.Connection connection;
+                try {
+                    connection = remoteClusterService.getConnection(clusterAlias);
+                } catch (NoSuchRemoteClusterException ex) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: Please see RemoteClusterAwareClient#doExecute.

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java [hunk 16]
```diff
@@ -218,7 +218,7 @@ public Transport.Connection getConnection(DiscoveryNode node, String cluster) {
      * Ensures that the given cluster alias is connected. If the cluster is connected this operation
      * will invoke the listener immediately.
      */
-    void ensureConnected(String clusterAlias, ActionListener<Void> listener) {
+    public void ensureConnected(String clusterAlias, ActionListener<Void> listener) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java]: I made an alternative to combine both ensureConnected and getConnection in https://github.com/elastic/elasticsearch/pull/95894/commits/56571d814be152a55feffcba1d42cfcb097d9b06. WDYT? Here we need to replicate partially the logic in RemoteClusterAwareClient#doExecute. Any suggestion are welcome.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 17]
```diff
@@ -746,8 +788,12 @@ static List<SearchShardIterator> getRemoteShardsIterator(
...
                     clusterAlias,
                     shardId,
-                    Arrays.asList(clusterSearchShardsGroup.getShards()),
-                    new OriginalIndices(finalIndices, originalIndices.indicesOptions())
+                    searchShardsGroup.allocatedNodes(),
+                    new OriginalIndices(finalIndices, originalIndices.indicesOptions()),
+                    null,
+                    null,
+                    searchShardsGroup.preFiltered(),
+                    searchShardsGroup.skipped()
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:778]: I know this is controversial, but the BWC layer is very small. I opened https://github.com/elastic/elasticsearch/pull/96123 to remove the `prefilter` flag. I will try to come up with an  alternative here.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 18]
```diff
@@ -582,50 +596,99 @@ static SearchResponseMerger createSearchResponseMerger(
...
+                } catch (NoSuchRemoteClusterException ex) {
+                    // attempt to reconnect but don't wait
+                    if (skipUnavailable) {
+                        remoteClusterService.ensureConnected(clusterAlias, ActionListener.noop());
+                    }
+                    throw ex;
+                }
+                final String[] indices = entry.getValue().indices();
+                // TODO: support point-in-time
+                if (searchContext == null && connection.getTransportVersion().onOrAfter(TransportVersion.V_8_500_000)) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: CCSDuelIT and ccs-rolling-upgrade-remote-cluster should cover this well already.

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import static org.hamcrest.Matchers.in;
+
+public class CCSCanMatchIT extends AbstractMultiClustersTestCase {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CCSCanMatchIT.java:52]: I believe an IT test here is more suitable because we are performing a real CCS search and assert our expectation. I've hardened this test (in https://github.com/elastic/elasticsearch/pull/95894/commits/0409f7300bb774fb1cbe7dce452611e2d32161d0) to cover all the bullets.

**Comment by dnhatn** [comment]
@javanna Thanks for reviews. I think I have addressed all your comments except the `preFiltered` flag. Can you take another look?

File: server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java]: It's ok for now, it's something I would like to clean up as a follow-up.

File: server/src/main/java/org/elasticsearch/action/search/SearchShardsResponse.java (hunk 6 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchShardsResponse.java:98]: it looks to me like this could be package private.

File: server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java (hunk 7 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java:148]: this question is unanswered, maybe the answer is yes :)

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java]: I see thanks for trying that. I think that I am still missing why we can't use the remote cluster client like we did before, sorry if that's a silly question but it's not so obvious to me :)

**Review by javanna — APPROVED** [review]
still a couple of questions but no blockers.

File: server/src/main/java/org/elasticsearch/action/search/SearchShardsResponse.java (hunk 6 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/SearchShardsResponse.java]: Fixed in https://github.com/elastic/elasticsearch/pull/95894/commits/8f5b715aaeac886f0d4f1551141ea5b0d067921b

File: server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java (hunk 7 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java]: Good catch. I've removed the old action in https://github.com/elastic/elasticsearch/pull/95894/commits/b3803e12c7811504c656046ba03923ec7324b7b9.

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java]: The reason that we can't use the remote cluster client is that we need to know the version the of the transport connection (between the local and remote clusters) before we decide if we can go with the new action.
